include/oneapi/dpl/internal/dynamic_selection_impl/dynamic_load_policy.h

```diff
@@ -0,0 +1,168 @@
+
+#include <atomic>
+#include <memory>
+#include <limits>
+#include <mutex>
+#include <vector>
+#include <exception>
+#include <type_traits>
+#if _DS_BACKEND_SYCL != 0
+    #include "oneapi/dpl/internal/dynamic_selection_impl/sycl_backend.h"
+#endif
+
+namespace oneapi {
+namespace dpl{
+namespace experimental{
+
+#if _DS_BACKEND_SYCL != 0
+  template <typename Backend=sycl_backend>
+#else
+  template <typename Backend>
+#endif
+  struct dynamic_load_policy {
+    private:
+    using backend_t = Backend;
+    using load_t = int;
+    using execution_resource_t = typename backend_t::execution_resource_t;
+
+    public:
+    //Policy Traits
+    using resource_type = typename backend_t::resource_type;
+    using wait_type = typename backend_t::wait_type;
+
+    //private:
+    struct resource_t {
+      execution_resource_t e_;
+      std::atomic<load_t> load_;
+      resource_t(execution_resource_t e) : e_(e) {load_=0;}
+    };
+    using resource_container_t = std::vector<std::shared_ptr<resource_t>>;
+
+    template<typename Policy>
+    class dl_selection_handle_t {
+      Policy policy_;
+      std::shared_ptr<resource_t> resource_;
+
+    public:
+      dl_selection_handle_t(const Policy& p,std::shared_ptr<resource_t> r )
+        : policy_(p), resource_(r) {}
+
+      auto unwrap() { return ::oneapi::dpl::experimental::unwrap(resource_->e_); }
+
+      Policy get_policy() { return policy_; };
+
+      void report(const execution_info::task_submission_t&) {
+        resource_->load_.fetch_add(1);
+      }
+
+      void report(const execution_info::task_completion_t&) const  {
+        resource_->load_.fetch_sub(1);
+      }
+    };
+
+    std::shared_ptr<backend_t> backend_;
+
+    using selection_type = dl_selection_handle_t<dynamic_load_policy<Backend>>;
+
+    struct state_t{
+        resource_container_t resources_;
+        int offset;
+        std::mutex m_;
+    };
+
+    std::shared_ptr<state_t> state_;
+
+    void initialize(int offset=0){
+        if(!state_){
+            backend_ = std::make_shared<backend_t>();
+            state_= std::make_shared<state_t>();
+            state_->offset=offset;
+            auto u =  get_resources();
+            for(auto x : u){
+              state_->resources_.push_back(std::make_shared<resource_t>(x));
+            }
+        }
+
+    }
+    void initialize(const std::vector<resource_type> &u, int offset=0) {
+        if(!state_){
+            backend_ = std::make_shared<backend_t>(u);
+            state_= std::make_shared<state_t>();
+            state_->offset=offset;
+            for(auto x : u){
+              state_->resources_.push_back(std::make_shared<resource_t>(x));
+            }
+        }
+    }
+
+    dynamic_load_policy(int offset=0) {
+        if(offset != deferred_initialization){
+            initialize(offset);
+        }
+    }
+
+    dynamic_load_policy(const std::vector<resource_type>& u,int offset=0) {
+        if(offset != deferred_initialization){
```

Comment:
The design looks weird to me. There are lots of questions to offset:
- Why do we have the same parameter for offset and deferred initialization?
- How initialization may be deferred if resources are passed?
- What's the point of having offset for the constructor without resources?
- Do we really need offset at all? The only use-case I could imagine is "the same set of resources multiple policy objects are constructed from" and you have different "start" point...

Overall, the same or similar questions are applicable to other policies as well

Reply:
Fixed. Offset now only remains for fixed_resource_policy. All policies use a special constructor that receives struct to indicate deferred initialization.